android/src/main/java/com/revenuecat/purchases_flutter/PurchasesFlutterPlugin.java

```diff
@@ -56,12 +72,51 @@ public boolean onViewDestroy(FlutterNativeView flutterNativeView) {
         });
     }
 
-    /**
-     * Plugin registration.
-     */
-    public static void registerWith(Registrar registrar) {
-        final MethodChannel channel = new MethodChannel(registrar.messenger(), "purchases_flutter");
-        channel.setMethodCallHandler(new PurchasesFlutterPlugin(registrar, channel));
+    @Override
+    public void onAttachedToEngine(@NonNull FlutterPluginBinding binding) {
+        onAttachedToEngine(binding.getBinaryMessenger(), binding.getApplicationContext());
+    }
+
+    private void onAttachedToEngine(BinaryMessenger messenger, Context applicationContext) {
+        this.channel = new MethodChannel(messenger, "purchases_flutter");
+        this.applicationContext = applicationContext;
+        this.channel.setMethodCallHandler(this);
+    }
+
+    @Override
+    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
+        try {
+            Purchases.getSharedInstance().close();
+        } catch (UninitializedPropertyAccessException e) {
+            // there's no instance so all good
+        }
+        channel.setMethodCallHandler(null);
+        this.channel = null;
+        this.applicationContext = null;
+    }
+
+    @Override
+    public void onAttachedToActivity(@NonNull ActivityPluginBinding binding) {
+        this.activity = binding.getActivity();
+    }
+
+    @Override
+    public void onReattachedToActivityForConfigChanges(@NonNull ActivityPluginBinding binding) {
+        onAttachedToActivity(binding);
+    }
+
+    @Override
+    public void onDetachedFromActivity() {
+        this.activity = null;
+    }
+
+    @Override
+    public void onDetachedFromActivityForConfigChanges() {
+        onDetachedFromActivity();
+    }
+
+    public Activity getActivity() {
+        return registrar != null ? registrar.activity() : activity;
```

Comment:
This is how they do in https://github.com/flutter/plugins/blob/88e85c6a48063cfab9111689179e2582a99b77ff/packages/google_sign_in/google_sign_in/android/src/main/java/io/flutter/plugins/googlesignin/GoogleSignInPlugin.java